File: puzzle/yweiyst/solution/client-code/js/game_geom.js
import { CubeFace, Matrix3x3, CardinalDirection } from './geom.js';

class KtaneGameGeom {
	// get player view relative to the front face
	static getRelativePlayerView(face) {
		let orient = Matrix3x3.IDENTITY;
		switch (face) {
		case CubeFace.FRONT:
			break;
		case CubeFace.BACK:
			orient = orient.leftMult(CubeFace.toRot(CubeFace.TOP));
			orient = orient.leftMult(CubeFace.toRot(CubeFace.TOP));
			break;
		case CubeFace.LEFT:
			orient = orient.leftMult(CubeFace.toRot(CubeFace.BOTTOM));
			break;
		case CubeFace.RIGHT:
			orient = orient.leftMult(CubeFace.toRot(CubeFace.TOP));
			break;
		case CubeFace.TOP:
			orient = orient.leftMult(CubeFace.toRot(CubeFace.LEFT));
			break;
		case CubeFace.BOTTOM:
			orient = orient.leftMult(CubeFace.toRot(CubeFace.RIGHT));
			break;
		default:
			console.error('invalild CubeFace');
			break;
		}
		return orient;
	}
	// returns an orientation that would display the face upright
	// same as relative player view since players should initially
	// see the face in upright orientation
	static getUprightOrientation(face) {
		return KtaneGameGeom.getRelativePlayerView(face);
	}
	static getPlayerViewMatrix(viewMat, playerSlot) {
		const relViewMat = KtaneGameGeom.getRelativePlayerView(playerSlot);
		return relViewMat.leftMult(viewMat);
	}
	static getActiveFace(viewMat, playerSlot) {
		const playerViewMat = KtaneGameGeom.getPlayerViewMatrix(viewMat, playerSlot);
		const activeVec = playerViewMat.multVec(CubeFace.FRONT_VEC);
		const activeFace = CubeFace.fromVec(activeVec);
		return activeFace;
	}
	// returns the direction the player's face is oriented relatve
	// to the default upright orientation
	static getPlayerRot(viewMat, playerSlot) {
		const playerViewMat = KtaneGameGeom.getPlayerViewMatrix(viewMat, playerSlot);
		const activeVec = playerViewMat.multVec(CubeFace.FRONT_VEC);
		const activeFace = CubeFace.fromVec(activeVec);
		let refMat = null;
		for (let i = 0; i < 4; i++) {
			refMat = (i == 0) ?
				KtaneGameGeom.getUprightOrientation(activeFace) :
				refMat.leftMult(CubeFace.toRot(activeFace));
			if (playerViewMat.equals(refMat)) {
				return CardinalDirection.CLOCKWISE_FROM_N[i];
			}
		}
		console.error('could not find orientation');
		return null;
	}
	static getRotFace(oldViewMat, newViewMat) {
		for (let i = 0; i < CubeFace.NUM; i++) {
			const refMat = oldViewMat.leftMult(CubeFace.toRot(i));
			if (refMat.equals(newViewMat)) {
				return i;
			}
		}
		console.error('could not find orientation');
		return null;
	}
	static rotDirToFace(d) {
		switch (d) {
		case CardinalDirection.N: return CubeFace.LEFT;
		case CardinalDirection.E: return CubeFace.TOP;
		case CardinalDirection.S: return CubeFace.RIGHT;
		case CardinalDirection.W: return CubeFace.BOTTOM;
		default:
			console.error('invalid CardinalDirection');
			return null;
		}
	}
	static rotByFace(viewMatrix, rotFace) {
		const rotAxis = CubeFace.toVec(rotFace);
		const rotMat = Matrix3x3.makeRotate90DegOrthoAA(rotAxis);
		const newViewMatrix = viewMatrix.leftMult(rotMat);
		return newViewMatrix;
	}
	static playerDirToRotFace(viewMatrix, playerFace, rotDir) {
		const playerViewMatrix = KtaneGameGeom.getPlayerViewMatrix(
			viewMatrix, playerFace
		);
		const rotFaceRel = KtaneGameGeom.rotDirToFace(rotDir);
		const rotAxisRel = CubeFace.toVec(rotFaceRel);
		const rotAxis = playerViewMatrix.multVec(rotAxisRel);
		const rotFace = CubeFace.fromVec(rotAxis);
		return rotFace;
	}

};

export { KtaneGameGeom };

File: hunt/special_puzzles/ktane/game_geom.py
from hunt.special_puzzles.ktane.geom import *

class KtaneGameGeom:
    @staticmethod
    def get_relative_player_view(face):
        orient = Matrix3x3.make_identity()
        if (face == CubeFace.back):
            orient = orient.left_mult(CubeFace.to_rot(CubeFace.top))
            orient = orient.left_mult(CubeFace.to_rot(CubeFace.top))
        elif (face == CubeFace.left):
            orient = orient.left_mult(CubeFace.to_rot(CubeFace.bottom))
        elif (face == CubeFace.right):
            orient = orient.left_mult(CubeFace.to_rot(CubeFace.top))
        elif (face == CubeFace.top):
            orient = orient.left_mult(CubeFace.to_rot(CubeFace.left))
        elif (face == CubeFace.bottom):
            orient = orient.left_mult(CubeFace.to_rot(CubeFace.right))
        return orient

    @staticmethod
    def get_player_view_matrix(view_mat, player_slot):
        rel_view_mat = KtaneGameGeom.get_relative_player_view(player_slot)
        return rel_view_mat.left_mult(view_mat)

    # gets the face to rotate if rotation is issued from front face
    @staticmethod
    def rot_dir_to_face(d):
        return {
            CardinalDirection.n: CubeFace.left,
            CardinalDirection.e: CubeFace.top,
            CardinalDirection.s: CubeFace.right,
            CardinalDirection.w: CubeFace.bottom
        }[d]

    @staticmethod
    def rot_by_face(view_matrix, rot_face):
        rot_axis = rot_face.get_vec()
        rot_mat = Matrix3x3.make_rotate_90deg_ortho_aa(rot_axis)
        new_view_matrix = view_matrix.left_mult(rot_mat)
        return new_view_matrix

    @staticmethod
    def player_dir_to_rot_face(view_matrix, player_face, rot_dir):
        player_view_matrix = KtaneGameGeom.get_player_view_matrix(
            view_matrix, player_face
        )
        rot_face_rel = KtaneGameGeom.rot_dir_to_face(rot_dir)
        rot_axis_rel = rot_face_rel.get_vec()
        rot_axis = player_view_matrix.mult_vec(rot_axis_rel)
        rot_face = CubeFace.from_vec(rot_axis)
        return rot_face

    @staticmethod
    def player_face_to_face(view_matrix, player_face):
        player_vec = player_face.get_vec()
        vec = view_matrix.mult_vec(player_vec)
        return CubeFace.from_vec(vec)

    @staticmethod
    def char_to_cube_face(c):
        return {
            'C': CubeFace.top,
            'G': CubeFace.bottom,
            'N': CubeFace.back,
            'E': CubeFace.right,
            'S': CubeFace.front,
            'W': CubeFace.left,
        }[c]
